File: java/src/main/java/P058.java
import java.util.*;

@SuppressWarnings("unused")
public class P058 implements Puzzle {

    @Override
    public String solve() {
        final BitSet primes = Util.primesBits(1_000_000_000);

        int n = 1;
        int primeCount = 0;

        for (int i = 0; ; ++i) {
            final int inc = (i / 4 + 1) * 2;
            final double primeRatio = (double) primeCount / (i + 1);

            if (primeRatio < 0.1 && inc > 3 && i % 4 == 0) {
                return "" + (inc - 1);
            }
            if (primes.get(n)) {
                ++primeCount;
            }
            n += inc;
        }
    }
}

File: erlang/src/p058.erl
-module(p058).

-export([start/0]).


f() ->
  f(1, 0, 0).

f(N, I, PrimeCount)  ->
  Inc = ((I div 4) + 1) * 2,
  PrimeRatio = PrimeCount / (I + 1),
  IsPrime = utils:is_prime(N),

  if
    Inc > 3 andalso I rem 4 == 0 andalso PrimeRatio < 0.1 -> Inc - 1;
    IsPrime -> f(N + Inc, I + 1, PrimeCount + 1);
    true    -> f(N + Inc, I + 1, PrimeCount)
  end.

start() ->
  io:format("~p ~n", [f()]).

